change min/max to overall_min/overall_max + update comparison results publisher

OVI3D0 opened the pull request:
### Description
This change updates min/max values in aggregated results to reflect the min/max values across all test executions, rather than just an average of the min/max values. I also changed the names from `min` and `max` to `overall_min` and `overall_max` for clarity. Changes were also made to the comparison results publisher class, so if an aggregated result with these new metrics were to be used in a comparison, these new 'overall' min and max values can still be compared without issue.

### Issues Resolved
#684 

### Testing
- [x] New functionality includes testing

`make test`

---
By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

gkamat commented on osbenchmark/aggregator.py:
```diff
@@ -202,15 +202,22 @@ def calculate_weighted_average(self, task_metrics: Dict[str, List[Any]], iterati
+                            weighted_metrics[metric]['overall_max'] = max(item_values)
+                        elif item_key == 'median':
+                            weighted_sum = sum(value * iterations for value in item_values)
+                            total_iterations = iterations * len(item_values)
```
Isn't this already available from above and does not need to be re-computed?

OVI3D0 replied:
True, I went ahead and moved this calculation outside of the loop.

gkamat commented on osbenchmark/aggregator.py:
```diff
@@ -202,15 +202,22 @@ def calculate_weighted_average(self, task_metrics: Dict[str, List[Any]], iterati
+                            weighted_sum = sum(value * iterations for value in item_values)
+                            total_iterations = iterations * len(item_values)
+                            weighted_metrics[metric][item_key] = weighted_sum / total_iterations
+                        else:
+                            weighted_sum = sum(value * iterations for value in item_values)
+                            total_iterations = iterations * len(item_values)
+                            weighted_metrics[metric][item_key] = weighted_sum / total_iterations
```
The `elif` is probably not needed here, since the code in the two arms is the same?

IanHoang replied:
+1 we can just have a comment in the else statement right after `else` stating what cases this for
```
else:
    # for items like median
```

OVI3D0 replied:
Got it, I added a comment to clarify what cases this is for

gkamat commented on osbenchmark/results_publisher.py:
```diff
@@ -464,16 +464,16 @@ def _write_results(self, metrics_table, metrics_table_console):
 
     def _publish_throughput(self, baseline_stats, contender_stats, task):
-        b_min = baseline_stats.metrics(task)["throughput"]["min"]
+        b_min = baseline_stats.metrics(task)["throughput"].get("overall_min") or baseline_stats.metrics(task)["throughput"]["min"]
```
It would probably be better to set the `overall_min` key prior, so dealing with a special case won't be necessary.

OVI3D0 replied:
I think this logic is necessary since the results publisher is also used for normal test executions which will not contain 'overall' min/max values

gkamat reviewed (changes requested):
Please consider adding (or updating) a test as well.

IanHoang commented on osbenchmark/results_publisher.py:
```diff
@@ -464,16 +464,16 @@ def _write_results(self, metrics_table, metrics_table_console):
         b_unit = baseline_stats.metrics(task)["throughput"]["unit"]
 
-        c_min = contender_stats.metrics(task)["throughput"]["min"]
+        c_min = contender_stats.metrics(task)["throughput"].get("overall_min") or contender_stats.metrics(task)["throughput"]["min"]
```
This change was made in the ComparisonResultsPublisher but should they also be made in the SummaryResultsPublisher?

OVI3D0 replied:
I don't think so, since the summary results publisher isn't used by the aggregator class. Although if we wanted to publish results after aggregation then this would be necessary

IanHoang reviewed (commented):
Could you add tests to results_publisher to confirm the changes work?

IanHoang reviewed (changes requested):
Left a few comments

OVI3D0 commented:
> Could you add tests to results_publisher to confirm the changes work?

I added a test for this, but let me know if I should add more!

I also updated the unit tests for the aggregator class.
